plugins/interactive/shellserver.go

```diff
@@ -0,0 +1,192 @@
+		log:       log,
+	}
+	s.c.L = &s.m
+	return s
+}
+
+// Wait will wait for all active shells to be done and return
+func (s *ShellServer) Wait() {
+	s.m.Lock()
+	defer s.m.Unlock()
+	for s.refCount > 0 {
+		s.c.Wait()
+	}
+}
+
+// WaitAndClose will wait for all active shells to be done and abort creation
+// of all new shells atomically.
+func (s *ShellServer) WaitAndClose() {
+	s.m.Lock()
+	defer s.m.Unlock()
+	for s.refCount > 0 {
+		s.c.Wait()
+	}
+
+	// Close done, aborting creation of new shells
+	select {
+	case <-s.done:
+	default:
+		close(s.done)
+	}
+}
+
+// Abort will abort all active interactive shells
+func (s *ShellServer) Abort() {
+	s.m.Lock()
+	defer s.m.Unlock()
+
+	select {
+	case <-s.done:
+	default:
+		close(s.done)
+	}
+}
+
+var upgrader = websocket.Upgrader{
+	HandshakeTimeout: ShellHandshakeTimeout,
+	ReadBufferSize:   ShellMaxMessageSize,
+	WriteBufferSize:  ShellMaxMessageSize,
+}
+
+func (s *ShellServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Quickly check that server haven't been aborted yet
+	select {
+	case <-s.done:
+		w.WriteHeader(http.StatusGone)
+		return
+	default:
+	}
+
+	// Get command and tty from query-string
+	qs := r.URL.Query()
+	command := qs["command"]
+	tty := strings.ToLower(qs.Get("tty")) == "true"
+
+	// Create a new shell, do this before we upgrade so we can return 410 on error
+	shell, err := s.makeShell(command, tty)
+	if err == engines.ErrSandboxTerminated || err == engines.ErrSandboxAborted {
+		w.WriteHeader(http.StatusGone)
+		return
+	}
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
+		debug("Failed to create shell, error: %s", err)
+		return
+	}
+
+	// Upgrade request to a websocket, abort the shell if upgrade fails
+	ws, err := upgrader.Upgrade(w, r, nil)
+	if err != nil {
+		debug("Failed to upgrade request to websocket, error: %s", err)
+		shell.Abort()
+		return
+	}
+
+	go s.handleShell(ws, shell)
+}
+
+func (s *ShellServer) handleShell(ws *websocket.Conn, shell engines.Shell) {
+	done := make(chan struct{})
+
+	// Create a shell handler
+	s.updateRefCount(1)
+	handler := NewShellHandler(ws, s.log.WithField("shell-instance-id", s.nextID()))
+
+	// Connect pipes
+	go ioext.CopyAndClose(shell.StdinPipe(), handler.StdinPipe())
+	go ioext.CopyAndClose(handler.StdoutPipe(), shell.StdoutPipe())
+	go ioext.CopyAndClose(handler.StderrPipe(), shell.StderrPipe())
+
+	// Start streaming
+	handler.Communicate(shell.SetSize, shell.Abort)
+
+	// Wait for call to abort all shells
+	go func() {
+		select {
+		case <-s.done:
+			shell.Abort()
+		case <-done:
+		}
+	}()
+
+	// Wait for the shell to terminate
+	success, _ := shell.Wait()
+	handler.Terminated(success)
+	s.updateRefCount(-1)
+
+	// Close done so we stop waiting for abort on all shells
+	select {
+	case <-done:
+	default:
+		close(done)
+	}
+}
+
+func (s *ShellServer) updateRefCount(change int) {
+	s.m.Lock()
+	s.refCount += change
+	if s.refCount <= 0 {
+		s.c.Broadcast()
+	}
+	s.m.Unlock()
```

Comment:
Doesn't this deadlock with `Wait` and `WaitAndClose`? If we call `Wait` with ```refCount > 0```, then `Wait` will wait forever the value reaches zero, and `updateRefCount` will wait forever the mutex be released.